routing/hopiterator.go

```diff
@@ -0,0 +1,168 @@
+package routing
+
+import (
+	"bytes"
+	"encoding/hex"
+	"github.com/btcsuite/golangcrypto/ripemd160"
+	"github.com/go-errors/errors"
+	"github.com/lightningnetwork/lightning-onion"
+	"github.com/roasbeef/btcd/btcec"
+	"github.com/roasbeef/btcutil"
+)
+
+// HopID represents the id which is used by propagation subsystem in order to
+// identify lightning network node.
+type HopID [ripemd160.Size]byte
+
+// NewHopID...
+func NewHopID(pubKey []byte) *HopID {
+	var routeId HopID
+	copy(routeId[:], btcutil.Hash160(pubKey))
+	return &routeId
+}
+
+func (id *HopID) String() string {
+	return hex.EncodeToString(id[:])
+}
+
+func (first *HopID) Equal(second *HopID) bool {
+	return bytes.Equal(first[:], second[:])
+}
+
+// HopIterator interface represent the entity which is able to give route
+// hops one by one. This interface is used to have an abstraction over the
+// algorithm which we use to determine the next hope in htlc route and also
+// help the unit test to create mock representation of such algorithm.
+type HopIterator interface {
+	// Next...
+	Next() *HopID
+
+	// ToBytes...
+	ToBytes() ([]byte, error)
+}
+
+// SphinxHopIterator...
+type SphinxHopIterator struct {
+	onionPacket  *sphinx.OnionPacket
+	sphinxPacket *sphinx.ProcessedPacket
+}
+
+var _ HopIterator = (*SphinxHopIterator)(nil)
+
+// NewSphinxHopIterator...
+func NewSphinxHopIterator(route *Route,
```

Comment:
This will have to be modified once the PR is rebased as the routing package has modified the usage.